pkg/gatherers/clusterconfig/node_logs.go

```diff
@@ -0,0 +1,149 @@
+//
+// * Location in archive: config/nodes/logs/
+// * See: docs/insights-archive-sample/config/nodes/logs
+// * Id in config: node_logs
+func (g *Gatherer) GatherNodeLogs(ctx context.Context) ([]record.Record, []error) {
+	clientSet, err := kubernetes.NewForConfig(g.gatherProtoKubeConfig)
+	if err != nil {
+		return nil, []error{err}
+	}
+	return gatherNodeLogs(ctx, clientSet.CoreV1())
+}
+
+func gatherNodeLogs(ctx context.Context, client corev1client.CoreV1Interface) ([]record.Record, []error) {
+	nodes, err := client.Nodes().List(ctx, metav1.ListOptions{LabelSelector: "node-role.kubernetes.io/master"})
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	records, errs := nodeLogRecords(client.RESTClient(), nodes)
+	return records, errs
+}
+
+// nodeLogRecords generate the records and errors list
+func nodeLogRecords(restClient rest.Interface, nodes *corev1.NodeList) ([]record.Record, []error) {
+	var errs []error
+	records := make([]record.Record, 0)
+
+	bufferSize := recorder.MaxArchiveSize * logCompressionRatio / len(nodes.Items) / 2
+	klog.V(2).Infof("Maximum buffer size: %v bytes", bufferSize)
+	buf := bytes.NewBuffer(make([]byte, 0, bufferSize))
+
+	for i := range nodes.Items {
+		name := nodes.Items[i].Name
+		uri := nodeLogResourceURI(restClient, name)
+		req := requestNodeLog(restClient, uri, logNodeMaxTailLines, logNodeUnit)
+
+		logString, err := nodeLogString(req, buf, bufferSize)
+		if err != nil {
+			klog.V(2).Infof("Error: %q", err)
+			errs = append(errs, err)
+		}
+
+		records = append(records, record.Record{
+			Name: fmt.Sprintf("config/node/logs/%s.log", name),
+			Item: marshal.Raw{Str: logString},
+		})
+	}
+
+	return records, errs
+}
+
+// nodeLogResourceURI creates the resource path URI to be fetched
+func nodeLogResourceURI(client rest.Interface, name string) string {
+	return client.Get().
+		Name(name).
+		Resource("nodes").SubResource("proxy", "logs").
+		Suffix("journal").URL().Path
+}
+
+// requestNodeLog creates the request to the API to retrieve the resource stream
+func requestNodeLog(client rest.Interface, uri string, tail int, unit string) *rest.Request {
+	return client.Get().RequestURI(uri).
+		SetHeader("Accept", "text/plain, */*").
+		SetHeader("Accept-Encoding", "gzip").
+		Param("tail", strconv.Itoa(tail)).
+		Param("unit", unit)
+}
+
+// nodeLogString retrieve the data from the stream, decompress it (if necessary) and return the string
+func nodeLogString(req *rest.Request, out *bytes.Buffer, size int) (string, error) {
+	in, err := req.Stream(context.TODO())
+	if err != nil {
+		return "", err
+	}
+	defer in.Close()
+
+	buf := bufio.NewReaderSize(in, size)
+	head, err := buf.Peek(1024)
+	if err != nil && err != io.EOF {
+		return "", err
+	}
+
+	_, err = gzip.NewReader(bytes.NewBuffer(head))
+	if err != nil {
+		// not gzipped stream
+		_, err = io.Copy(out, buf)
+		if err != nil && err != io.EOF {
+			return "", err
+		}
+
+		return out.String(), nil
```

Comment:
So if it's not gzip stream then we return only first 1024 bytes? It seems a bit inconsistent comparing to filtering below.

Reply:
Nope, it will not. The returned buffer has the same size for both cases, gzipped or un-gzipped. The peek is only used to validate if the stream is gzipped or not. It basically the same logic that the `oc` uses to get it done.